### PR TITLE
Knife tool with 'multi-source-editing'

### LINKED ISSUE #190: Add knife tool
Relates to #9 and #30

### PR DESCRIPTION
**Pull request by ollimeier**
Fixes #190

I have tested it extensively and I have the feel that it does what it is supposed to do. However, I still keep this PR as 'draft', because there are quite a few comments we may or may not want to keep.

### CODE REVIEW

**Review by justvanrossum — COMMENTED** [review]
I'm finding many cases that don't work well, as well as cases where compatibility gets broken. 

I feel we're not on the right track, and this is not heading towards a solid solution :(

Also, I don't like the changes at the end of insertPoint to compute the inserted point indices, and I'm pretty sure that can be solved in a better way.

**Comment by ollimeier** [comment]
> I'm finding many cases that don't work well, as well as cases where compatibility gets broken.
> 
> I feel we're not on the right track, and this is not heading towards a solid solution :(
> 
> Also, I don't like the changes at the end of insertPoint to compute the inserted point indices, and I'm pretty sure that can be solved in a better way.

Thank you for your honest feedback. I much appreciate it and I am very sorry, that this not heading towards the right solution. 
It would help me a lot if you could give me more specific instructions, because your feedback is unfortunately a bit vague. 
For example, it would help me a lot if you could give me examples of where it doesn't work and what the expectations are (how it should work).
Or do you have any other suggestions on how we should proceed?

**Comment by justvanrossum** [comment]
I added some initial test for insertPoint and splitPathAtPointIndices in test-path-functions.js, in #1569. It would be good to rebase and perhaps see if the insertPoint tests can be enhanced to cover your changes.

**Comment by justvanrossum** [comment]
> It would be good to rebase

I did the rebase and tweaked the insertPoint test (but didn't yet add test cases for multi-t intersections).

**Comment by ollimeier** [comment]
> > It would be good to rebase
> 
> I did the rebase and tweaked the insertPoint test (but didn't yet add test cases for multi-t intersections).

Awesome. I will have a look. Thanks.

**Comment by ollimeier** [comment]
@justvanrossum I took your idea of storing useful information like `isClosed` or `winding` to improve the knife tool. Here is an example how only closed contours will be reconnected and open contours will be split only:

https://github.com/user-attachments/assets/cefd22ee-aa1d-4126-b216-33d4adc81713

**Comment by ollimeier** [comment]
@justvanrossum I also improved the way how to find the right connection points via the winding direction, here an example we used before, but now with a much better result, am I right?


https://github.com/user-attachments/assets/cbf4ee11-c55b-4161-8179-ff6fd375306e

**Comment by ollimeier** [comment]
@justvanrossum This case shows a split (only), but I would expect closed contours:


https://github.com/user-attachments/assets/1d497cab-7184-4522-88cb-27a134da2c37


This is the glyph:

`{"layerGlyphs":[{"layerName":"RobotoAvar2-wght400/foreground","location":{},"glyph":{"xAdvance":1174,"path":{"coordinates":[259,438,259,580.0146751704024,343.2912375511347,775,510,775,541,775,570,769,596,758,622,747,643,734,660,720,707,720,645,412,637,371,640,342,654,326,669,309,690,301,717,301,766,301,805,326,835,375,866,424,881,488,881,567,881,660,849,742,784,813,719,884,627,919,510,919,262.54244700809903,919,96,705.5122992029173,96,469,96,245.94764521788545,227.5400573055149,81,459,81,507,81,548,86,581,95,614,104,643,115,666,127,666,47,641,32,613,20,580,12,547,4,507,0,459,0,186.52088313953038,0,0,199.81027398225842,0,469,0,749.8124208687354,223.03114736813416,1000,510,1000,648,1000,760,958,846,874,931,790,974,688,974,567,974,467,950,386,903,324,855,262,797,231,728,231,688,231,653,240,623,257,593,274,571,297,557,324,540,297,519,274,495,257,471,240,445,231,417,231,302.98591437486624,231,259,339.648018954083,358,438,358,381.8539037577926,373.6235282483433,312,443,312,467,312,489,321,509,339,528,356,543,388,552,433,604,681,593,688,581,694,567,698,554,701,540,703,526,703,398.91626395337977,703,358,538.1292834897911],"pointTypes":[8,2,2,8,2,2,8,2,2,0,0,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,0,0,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,0,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,2,2,8,0,2,2,0,2,2,8,2,2],"contourInfo":[{"endPoint":80,"isClosed":true},{"endPoint":99,"isClosed":true}],"pointAttributes":null},"components":[],"anchors":[],"guidelines":[]}}]}
`

**Comment by ollimeier** [comment]
@justvanrossum I am not sure if it's possible to find a solution for that, but it feels a bit unexpected:


https://github.com/user-attachments/assets/3e5a5f99-59d3-40cc-86fe-367833303b00



Glyph:
`{"layerGlyphs":[{"layerName":"RobotoAvar2-wght400/foreground","location":{},"glyph":{"xAdvance":1200,"path":{"coordinates":[170,0,170,1456,652,1456,892,1456,1150,1215,1150,1015,1150,815,892,570,652,570,266,570,266,728,652,728,822,728,958,895,958,1015,958,1135,822,1298,652,1298,362,1298,362,0],"pointTypes":[0,0,8,1,1,8,1,1,8,0,0,8,1,1,8,1,1,8,0,0],"contourInfo":[{"endPoint":19,"isClosed":true}],"pointAttributes":null},"components":[],"anchors":[],"guidelines":[]}}]}`

**Comment by ollimeier** [comment]
Same here: 


https://github.com/user-attachments/assets/5a54be24-80fc-4e7b-90bd-4a1c4b6d400f



Glyph:
{"layerGlyphs":[{"layerName":"RobotoAvar2-wght400/foreground","location":{},"glyph":{"xAdvance":1106,"path":{"coordinates":[450,-20,300,-20,90,150,90,310,90,470,344,648,540,648,871,648,871,516,560,516,420,516,276,410,276,320,276,230,380,128,490,128,600,128,778,280,778,365,778,720,777,815,680,934,560,934,440,934,337,835,337,770,337,765,152,765,152,770,152,895,375,1082,565,1082,755,1082,964,900,964,720,964,240,964,175,984,60,1002,16,1002,0,822,0,808,25,785,105,782,145,745,80,580,-20],"pointTypes":[8,1,1,8,1,1,8,0,0,8,1,1,8,1,1,8,1,1,0,8,1,1,8,1,1,8,0,0,8,1,1,8,1,1,8,8,1,1,0,0,0,1,1,0,1,1],"contourInfo":[{"endPoint":45,"isClosed":true}],"pointAttributes":null},"components":[],"anchors":[],"guidelines":[]}}]}

**Comment by ollimeier** [comment]
@justvanrossum  Another split, but no connection:

https://github.com/user-attachments/assets/454b33ce-a086-4cc0-92ff-b7e648dc8b7b



Glyph:
`{"layerGlyphs":[{"layerName":"RobotoAvar2-wght400/foreground","location":{},"glyph":{"xAdvance":1260,"path":{"coordinates":[170,0,170,1456,650,1456,870,1456,1122,1270,1122,1070,1122,950,1000,800,910,768,910,764,1020,732,1160,560,1160,420,1160,220,897,0,680,0,358,158,680,158,830,158,970,300,970,420,970,540,830,680,680,680,264,680,264,836,650,836,800,836,932,960,932,1070,932,1180,800,1298,650,1298,358,1298],"pointTypes":[0,0,8,1,1,8,1,1,0,0,1,1,8,1,1,8,0,8,1,1,8,1,1,8,0,0,8,1,1,8,1,1,8,0],"contourInfo":[{"endPoint":15,"isClosed":true},{"endPoint":33,"isClosed":true}],"pointAttributes":null},"components":[],"anchors":[],"guidelines":[]}}]}`

**Comment by ollimeier** [comment]
@justvanrossum I guess, that's it. Great job. Great new tool.

File: src/fontra/client/core/path-functions.js [hunk 1]
```diff
@@ -109,6 +151,18 @@ function impliedPoint(pointA, pointB) {
   };
 }
 
+function bezierSplitMultiple(bezier, ts) {
+  // it's possible to have 3 ts
+  ts = [0, ...ts, 1];
+  ts.sort((a, b) => a - b);
```
[the remark this answers is not in the document]
Comment (justvanrossum) [reply src/fontra/client/core/path-functions.js:157]: Ah yes, that sort can go.

File: src/fontra/views/editor/edit-tools-knife.js [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+    await this.sceneController.editGlyphAndRecordChanges(
+      (glyph) => {
+        const editLayerGlyphs = this.sceneController.getEditingLayerFromGlyphLayers(
+          glyph.layers
+        );
+        for (const layerGlyph of Object.values(editLayerGlyphs)) {
+          doCutLayerGlyph(intersections, sortedIntersections, layerGlyph.path);
+        }
+        return `Knife Tool cut`;
```
[the remark this answers is not in the document]
Comment (justvanrossum) [reply src/fontra/views/editor/edit-tools-knife.js]: Good catch, I renamed the function and the undo label.

File: src/fontra/views/editor/edit-tools-knife.js [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+      true
+    );
+  }
+
+  deactivate() {
+    this.canvasController.requestUpdate();
+  }
+}
+
+function getIntersections(glyphController, p1, p2) {
```
[the remark this answers is not in the document]
Comment (justvanrossum) [reply src/fontra/views/editor/edit-tools-knife.js]: I was wondering, too, and now that I've tried I agree that it's redundant.

**Pull request merged**